asserts/snapasserts/validation_sets.go

```diff
@@ -44,6 +60,57 @@ func (e *ValidationSetsConflictError) Error() string {
 	return buf.String()
 }
 
+// ValidationSetsValidationError describes an error arising
+// from validation of snaps against ValidationSets.
+type ValidationSetsValidationError struct {
+	// snapName -> validationSetKey
+	MissingSnaps       map[string][]string
+	InvalidSnaps       map[string][]string
+	WrongRevisionSnaps map[string]map[snap.Revision][]string
+	Sets               map[string]*asserts.ValidationSet
+}
+
+func (e *ValidationSetsValidationError) Error() string {
+	buf := bytes.NewBufferString("validation sets assertions are not met:")
+	printDetails := func(header string, details map[string][]string,
+		printSnap func(snapName string, keys []string) string) {
+		if len(details) == 0 {
+			return
+		}
+		fmt.Fprintf(buf, "\n- %s:", header)
+		for snapName, validationSetKeys := range details {
+			fmt.Fprintf(buf, "\n  - %s", printSnap(snapName, validationSetKeys))
+		}
+	}
+
+	printDetails("missing required snaps", e.MissingSnaps, func(snapName string, validationSetKeys []string) string {
+		return fmt.Sprintf("%s (required by sets %s)", snapName, strings.Join(validationSetKeys, ","))
+	})
+	printDetails("invalid snaps", e.InvalidSnaps, func(snapName string, validationSetKeys []string) string {
+		return fmt.Sprintf("%s (invalid for sets %s)", snapName, strings.Join(validationSetKeys, ","))
+	})
+
+	if len(e.WrongRevisionSnaps) > 0 {
+		fmt.Fprint(buf, "\n- snaps at wrong revisions:")
+		for snapName, revisions := range e.WrongRevisionSnaps {
+			revisionsSorted := make([]int, 0, len(revisions))
+			for rev := range revisions {
+				revisionsSorted = append(revisionsSorted, rev.N)
+			}
+			sort.Ints(revisionsSorted)
+			t := make([]string, 0, len(revisionsSorted))
+			for _, rev := range revisionsSorted {
+				r := snap.R(rev)
+				keys := revisions[r]
+				t = append(t, fmt.Sprintf("at revision %s by sets %s", r, strings.Join(keys, ",")))
```

Comment:
I suppose with some care we wouldn't need the []string t, but could append to buf directly ? not sure it's super important though

Reply:
Yes, I did manipulate buf initially but it seemed a little bit convoluted; this introduces a little bit of an overheard but IMHO is easier to follow.